cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,492 @@
+            elif my_status == 1: # found solution, but not optimality proven
+                assert self.xct_solver.hasSolution()
+                solsfound += 1
+                self.getSolAndObj()
+                self.xct_solver.invalidateLastSol()
+            elif my_status == 2: # found inconsistency
+                pass # TODO
+            elif my_status == 3: # found timeout
+                pass # TODO
+
+        return solsfound
+
+
+    def solver_var(self, cpm_var, encoding="onehot"):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+        if is_num(cpm_var):  # shortcut, eases posting constraints
+            return cpm_var
+
+        # variables to be translated should be positive
+        assert(not isinstance(cpm_var, NegBoolView))
+
+        # return it if it already exists
+        if cpm_var in self._varmap:
+            return self._varmap[cpm_var]
+
+        # create if it does not exist
+        revar = str(cpm_var)
+        if isinstance(cpm_var, _BoolVarImpl):
+            self.xct_solver.addVariable(revar,0,1)
+        elif isinstance(cpm_var, _IntVarImpl):
+            lb = cpm_var.lb
+            ub = cpm_var.ub
+            if max(abs(lb),abs(ub)) > 1e18:
+                # larger than 64 bit should be passed by string
+                self.xct_solver.addVariable(revar,str(lb), str(ub), encoding)
+            else:
+                self.xct_solver.addVariable(revar,lb,ub, encoding)
+        else:
+            raise NotImplementedError("Not a known var {}".format(cpm_var))
+        self._varmap[cpm_var] = revar
+        return revar
+    
+
+    def solver_var_no_num(self, cpm_var, encoding="onehot"):
+        assert not is_num(cpm_var)
+        return self.solver_var(cpm_var, encoding)
+
+
+    def objective(self, expr, minimize):
+        """
+            Post the given expression to the solver as objective to minimize/maximize
+
+            - expr: Expression, the CPMpy expression that represents the objective function
+            - minimize: Bool, whether it is a minimization problem (True) or maximization problem (False)
+
+            'objective()' can only be called once
+        """
+        assert not self.has_objective, "Exact accepts an objective function only once."
+        self.has_objective = True
+        self.objective_minimize = minimize
+
+        # make objective function non-nested
+        (flat_obj, flat_cons) = flatten_objective(expr)
+        self += flat_cons  # add potentially created constraints
+        self.user_vars.update(get_variables(flat_obj))  # add objvars to vars
+
+        # make objective function or variable and post
+        xct_coefs,xct_vars,xct_rhs = self._make_numexpr(flat_obj,0)
+        if not self.objective_minimize:
+            xct_coefs = [-x for x in xct_coefs]
+
+        self.xct_solver.init(xct_coefs,xct_vars)
+        self.solver_is_initialized = True
```

Comment:
I don't think this is the right place.

Also, what if I add an objective before a constraint? or no, it will actually call add, so add will always init the solver (and set this param)